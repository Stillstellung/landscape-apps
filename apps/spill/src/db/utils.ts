import {
  AnyPrimaryModel,
  PrimarySchemaName,
  Results,
  SchemaModel,
  SchemaName,
} from './types';

/**
 * Forcibly casts a realm results object to a plain, read-only array. Results and the
 * plain array are functionally identical in practice, but there's a subtlety
 * with scopeables that causes issues.
 *
 * Uncomment these lines if you want to check out the error:
 * let results: Realm.Results<any>;
 * let arrayResults = results as ReadonlyArray<any>;
 */
export function asReadOnlyArray<T extends SchemaName>(
  results: Results<T>,
): ReadonlyArray<SchemaModel<T>> {
  return results as unknown as ReadonlyArray<SchemaModel<T>>;
}

// Collection Proxy

const numericRegEx = /^-?\d+$/;

function getCacheKey(id: string) {
  return `${id}`;
}

/**
 * Ensure objects in a collection maintain referential equality.
 *
 * Mostly taken from here:
 * https://github.com/realm/realm-js/blob/main/packages/realm-react/src/cachedCollection.ts
 * Difference is that this version does not update when the query updates, and
 * therefore *shouldn't* (famous last words) need to be manually disposed.
 *
 * The reason that this is necessary is that objects returned from
 * `realm.objects` are proxied by default and do not have referential equality.
 * So, `objectA === objectA` will always return falso, wreaking havoc on  memoized
 * components. This is not the case for `useQuery` results, which are wrapped in
 * the above-mentioned `cachedCollection`.
 */
export const createCachingCollectionProxy = <T extends PrimarySchemaName>(
  collection: Results<T>,
): Results<T> => {
  const objectCache = new Map<string, SchemaModel<T>>();
  const cachedCollectionHandler: ProxyHandler<Results<T>> = {
    get: function (target, key, receiver) {
      // Pass functions through
      const value = Reflect.get(target, key, receiver);
      if (typeof value === 'function') {
        return value;
      }

      // If the key is not numeric, pass it through
      if (typeof key === 'symbol' || !numericRegEx.test(key)) {
        return value;
      }

      // If the key is numeric, check if we have a cached object for this key
      const index = Number(key);
      const object = target[index];

      // If the collection is modeled in a way that objects can be null
      // then we should return null instead of undefined to stay semantically
      // correct
      if (object === null) {
        return null;
      } else if (typeof object === 'undefined') {
        // If there is no object at this index, return undefined
        return undefined;
      }

      const objectId = getObjectId(object);
      const cacheKey = getCacheKey(objectId);

      // If we do, return it...
      if (objectCache.get(cacheKey)) {
        return objectCache.get(cacheKey);
      }

      // If not then this index has either not been accessed before, or has been invalidated due
      // to a modification. Fetch it from the collection and store it in the cache
      objectCache.set(cacheKey, object);

      return object;
    },
  };

  return new Proxy(collection, cachedCollectionHandler);
};

export const getObjectId = <T extends AnyPrimaryModel>(item: T) => {
  return item.id;
};
