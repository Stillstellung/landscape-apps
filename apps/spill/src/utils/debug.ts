import {useEffect, useMemo, useRef} from 'react';

// expands object types one level deep
export type Expand<T> = T extends infer O ? {[K in keyof O]: O[K]} : never;

// expands object types recursively
export type ExpandRecursively<T> = T extends object
  ? T extends infer O
    ? {[K in keyof O]: ExpandRecursively<O[K]>}
    : never
  : T;

const instances: Record<string, number> = {};
export const useInstanceId = (prefix: string) => {
  instances[prefix] ||= 0;
  return useMemo(() => {
    return `${prefix}-${instances[prefix]++}`;
  }, [prefix]);
};

export function createLogger(
  tag: string,
  enabled = isDev(),
): Pick<Console, 'log' | 'warn' | 'error'> {
  const prefix = `[${tag}]`;
  const noop = () => {};
  const log = enabled ? console.log.bind(console, prefix) : noop;
  const warn = enabled ? console.warn.bind(console, prefix) : noop;
  const error = enabled ? console.error.bind(console, prefix) : noop;
  return {log, warn, error};
}

export function stopwatch(label: string) {
  const startTime = Date.now();
  return {
    stop() {
      const endTime = Date.now();
      console.log(`${label} took ${endTime - startTime}ms`);
    },
  };
}

const timers: Record<string, number> = {};
const timerLogger = createLogger('timer', true);

export const timer = {
  start(label: string) {
    timers[label] = Date.now();
  },
  stop(label: string) {
    const endTime = Date.now();
    const startTime = timers[label];
    startTime
      ? timerLogger.log(`${label} took ${endTime - startTime}ms`)
      : timerLogger.log('no timer found for', label);
  },
};

export function logDuration(label: string, cb: () => void) {
  const startTime = Date.now();
  const result = cb();
  const endTime = Date.now();
  console.log(`${formatTag(label)}took ${endTime - startTime}ms`);
  return result;
}

const formatTag = (tag: string) => `[${tag}]`;

export function isDev() {
  return __DEV__;
}

export function usePrevious<T>(value: T): T | undefined {
  const ref = useRef<T>();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}

export function useLogChange(label: string, value: any, logValue = false) {
  const prevValue = usePrevious(value);
  if (prevValue !== value) {
    if (logValue) {
      console.log(`${label}:change`, 'CURRENT', value, 'PREVIOUS', prevValue);
    } else {
      console.log(`${label}:change`);
    }
  }
}
