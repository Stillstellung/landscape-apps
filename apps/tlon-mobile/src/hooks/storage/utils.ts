import { createDevLogger } from '@tloncorp/shared/dist';
import { RNFile, Uploader } from '@tloncorp/shared/dist/api';

import type { ShipInfo } from '../../contexts/ship';
import storage from '../../lib/storage';

const logger = createDevLogger('storage utils', true);

const fetchImageFromUri = async (uri: string) => {
  try {
    logger.log('fetchImageFromUri', uri);
    const response = await fetch(uri);
    logger.log('fetchImageFromUri response', response);
    const blob = await response.blob();
    const name = uri.split('/').pop();

    const file: RNFile = {
      blob,
      name: name ?? 'channel-image',
      type: blob.type,
    };

    return file;
  } catch (e) {
    console.error(e);
  }
};

export const handleImagePicked = async (uri: string, uploader: Uploader) => {
  logger.log('handleImagePicked', uri, uploader);
  try {
    const image = await fetchImageFromUri(uri);
    if (!image) {
      logger.log('no image');
      return;
    }

    logger.log({ image });

    await uploader?.uploadFiles([image]);
  } catch (e) {
    console.error(e);
  }
};

const getShipInfo = async () => {
  const shipInfo = (await storage.load({ key: 'store' })) as
    | ShipInfo
    | undefined;

  return shipInfo;
};

export const getIsHosted = async () => {
  const shipInfo = await getShipInfo();
  const isHosted = shipInfo?.shipUrl?.endsWith('tlon.network');
  return isHosted;
};

export const getHostingUploadURL = async () => {
  const isHosted = await getIsHosted();
  return isHosted ? 'https://memex.tlon.network' : '';
};
