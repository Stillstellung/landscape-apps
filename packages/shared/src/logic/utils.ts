import { differenceInDays, endOfToday, format } from 'date-fns';
import emojiRegex from 'emoji-regex';
import _ from 'lodash';

type App = 'chat' | 'heap' | 'diary';

export function checkNest(nest: string) {
  if (nest.split('/').length !== 3) {
    console.error('Invalid nest:', nest);
  }
}

export function nestToFlag(nest: string): [App, string] {
  checkNest(nest);
  const [app, ...rest] = nest.split('/');

  return [app as App, rest.join('/')];
}

export function makePrettyTime(date: Date) {
  return format(date, 'HH:mm');
}

export interface DayTimeDisplay {
  original: Date;
  diff: number;
  day: string;
  time: string;
  asString: string;
}

export function makePrettyDayAndTime(date: Date): DayTimeDisplay {
  const diff = differenceInDays(endOfToday(), date);
  const time = makePrettyTime(date);
  let day = '';
  switch (true) {
    case diff === 0:
      day = 'Today';
      break;
    case diff === 1:
      day = 'Yesterday';
      break;
    case diff > 1 && diff < 8:
      day = format(date, 'cccc');
      break;
    default:
      day = `${format(date, 'LLLL')} ${format(date, 'do')}`;
  }

  return {
    original: date,
    diff,
    time,
    day,
    asString: `${day} • ${time}`,
  };
}

export interface DateDayTimeDisplay extends DayTimeDisplay {
  fullDate: string;
}

export function makePrettyDayAndDateAndTime(date: Date): DateDayTimeDisplay {
  const fullDate = `${format(date, 'LLLL')} ${format(date, 'do')}, ${format(
    date,
    'yyyy'
  )}`;
  const dayTime = makePrettyDayAndTime(date);

  if (dayTime.diff >= 8) {
    return {
      ...dayTime,
      fullDate,
      asString: `${fullDate} • ${dayTime.time}`,
    };
  }

  return {
    ...dayTime,
    fullDate,
    asString: `${dayTime.asString} • ${fullDate}`,
  };
}

export function isSingleEmoji(input: string): boolean {
  const regex = emojiRegex();
  const matches = input.match(regex);

  return (
    (matches &&
      matches.length === 1 &&
      matches.length === _.split(input, '').length) ??
    false
  );
}
