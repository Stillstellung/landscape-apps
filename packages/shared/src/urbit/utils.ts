import * as ub from './channel';
import * as ubc from './content';

type App = 'chat' | 'heap' | 'diary';

export function checkNest(nest: string) {
  if (nest.split('/').length !== 3) {
    console.error('Invalid nest:', nest);
  }
}

export function nestToFlag(nest: string): [App, string] {
  checkNest(nest);
  const [app, ...rest] = nest.split('/');

  return [app as App, rest.join('/')];
}

export function getChannelType(channelId: string) {
  const [app] = nestToFlag(channelId);

  if (app === 'chat') {
    return 'chat';
  } else if (app === 'heap') {
    return 'gallery';
  } else if (app === 'diary') {
    return 'notebook';
  } else {
    return 'chat';
  }
}

export function getTextContent(story?: ub.Story | undefined) {
  if (!story) {
    return;
  }
  return story
    .map((verse) => {
      if (ubc.isBlock(verse)) {
        return getBlockContent(verse.block);
      } else {
        return getInlinesContent(verse.inline);
      }
    })
    .filter((v) => !!v && v !== '')
    .join(' ')
    .trim();
}

export function getBlockContent(block: ub.Block) {
  if (ub.isImage(block)) {
    return '[image]';
  } else if (ub.isCite(block)) {
    return '[ref]';
  } else if (ubc.isHeader(block)) {
    return block.header.content.map(getInlineContent);
  } else if (ubc.isCode(block)) {
    return block.code.code;
  } else if (ubc.isListing(block)) {
    return getListingContent(block.listing);
  }
}

export function getListingContent(listing: ub.Listing): string {
  if (ubc.isListItem(listing)) {
    return listing.item.map(getInlineContent).join(' ');
  } else {
    return listing.list.items.map(getListingContent).join(' ');
  }
}

export function getInlinesContent(inlines: ubc.Inline[]): string {
  return inlines
    .map(getInlineContent)
    .filter((v) => v && v !== '')
    .join(' ');
}

export function getInlineContent(inline: ubc.Inline): string {
  if (ubc.isBold(inline)) {
    return inline.bold.map(getInlineContent).join(' ');
  } else if (ubc.isItalics(inline)) {
    return inline.italics.map(getInlineContent).join(' ');
  } else if (ubc.isLink(inline)) {
    return inline.link.content;
  } else if (ubc.isStrikethrough(inline)) {
    return inline.strike.map(getInlineContent).join(' ');
  } else if (ubc.isBlockquote(inline)) {
    return inline.blockquote.map(getInlineContent).join(' ');
  } else if (ubc.isInlineCode(inline)) {
    return inline['inline-code'];
  } else if (ubc.isBlockCode(inline)) {
    return inline.code;
  } else if (ubc.isBreak(inline)) {
    return '';
  } else if (ubc.isShip(inline)) {
    return inline.ship;
  } else if (ubc.isTag(inline)) {
    return inline.tag;
  } else if (ubc.isBlockReference(inline)) {
    return inline.block.text;
  } else if (ubc.isTask(inline)) {
    return inline.task.content.map(getInlineContent).join(' ');
  } else {
    return inline;
  }
}
