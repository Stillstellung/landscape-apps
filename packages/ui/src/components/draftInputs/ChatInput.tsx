import { SafeAreaView } from 'react-native-safe-area-context';

import { MessageInput } from '../MessageInput';
import { ParentAgnosticKeyboardAvoidingView } from '../ParentAgnosticKeyboardAvoidingView';
import { DraftInputContext } from './shared';

export function ChatInput({
  draftInputContext,
}: {
  draftInputContext: DraftInputContext;
}) {
  const {
    channel,
    clearDraft,
    editPost,
    editingPost,
    getDraft,
    group,
    send,
    setEditingPost,
    setShouldBlur,
    shouldBlur,
    storeDraft,
  } = draftInputContext;
  if (editingPost != null) {
    return null;
  }

  return (
    <SafeAreaView edges={['right', 'left', 'bottom']}>
      <ParentAgnosticKeyboardAvoidingView>
        <MessageInput
          shouldBlur={shouldBlur}
          setShouldBlur={setShouldBlur}
          send={send}
          channelId={channel.id}
          groupMembers={group?.members ?? []}
          storeDraft={storeDraft}
          clearDraft={clearDraft}
          getDraft={getDraft}
          editingPost={editingPost}
          setEditingPost={setEditingPost}
          editPost={editPost}
          channelType={channel.type}
          showInlineAttachments
          showAttachmentButton
        />
      </ParentAgnosticKeyboardAvoidingView>
    </SafeAreaView>
  );
}
