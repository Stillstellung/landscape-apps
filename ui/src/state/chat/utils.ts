import { PostEssay } from '@/types/channel';
import { Club, ReplyDelta, WritDeltaAdd } from '@/types/dms';
import { formatUd, unixToDa } from '@urbit/aura';

export default function emptyMultiDm(): Club {
  return {
    hive: [],
    team: [],
    meta: {
      title: '',
      description: '',
      image: '',
      cover: '',
    },
  };
}

function makeId() {
  const time = Date.now();
  return {
    id: `${window.our}/${formatUd(unixToDa(time))}`,
    time,
  };
}

export function createMessage(
  whom: string,
  mem: PostEssay,
  replying?: string
): { id: string; delta: WritDeltaAdd | ReplyDelta } {
  const { id, time } = makeId();
  const memo: Omit<PostEssay, 'kind-data'> = {
    content: mem.content,
    author: mem.author,
    sent: time,
  };

  let delta: WritDeltaAdd | ReplyDelta;
  if (!replying) {
    delta = {
      add: {
        memo,
        kind: null,
        time: null,
      },
    };
  } else {
    delta = {
      reply: {
        id,
        meta: null,
        delta: {
          add: {
            memo,
            time: null,
          },
        },
      },
    };
  }

  return { id, delta };
}
