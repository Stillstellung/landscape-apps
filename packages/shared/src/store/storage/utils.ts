import { preSig } from '@urbit/api';
import { deSig, formatDa, unixToDa } from '@urbit/aura';

import { RNFile, Uploader, client, scry } from '../../api';
import { createDevLogger } from '../../debug';

const logger = createDevLogger('storage utils', true);

const fetchImageFromUri = async (uri: string, base64: string) => {
  try {
    logger.log('fetchImageFromUri', uri);
    const response = await fetch(uri);
    logger.log('fetchImageFromUri response', response);
    const blob = await response.blob();
    const name = uri.split('/').pop();

    const file: RNFile = {
      blob,
      name: name ?? 'channel-image',
      type: blob.type,
    };

    return file;
  } catch (e) {
    console.error(e);
  }
};

export const handleImagePicked = async (
  uri: string,
  base64: string,
  uploader: Uploader
) => {
  logger.log('handleImagePicked', uri, uploader);
  try {
    const image = await fetchImageFromUri(uri, base64);
    if (!image) {
      logger.log('no image');
      return;
    }

    logger.log({ image });

    await uploader?.uploadFiles([image]);
  } catch (e) {
    console.error(e);
  }
};

export const getShipInfo = async () => {
  const { ship, url } = client;

  if (!ship) {
    return { ship: '', shipUrl: '' };
  }

  return { ship: preSig(ship), shipUrl: url };
};

export const getIsHosted = async () => {
  const shipInfo = await getShipInfo();
  const isHosted = shipInfo?.shipUrl?.endsWith('tlon.network');
  return isHosted;
};

export const getHostingUploadURL = async () => {
  const isHosted = await getIsHosted();
  return isHosted ? 'https://memex.tlon.network' : '';
};

export const getMemexUploadUrl = async (ship: string, fileName: string) => {
  const presignedUrl = 'https://memex.tlon.network';
  const key = `${ship}/${deSig(formatDa(unixToDa(new Date().getTime())))}-${fileName.split(' ').join('-')}`;
  const url = `${presignedUrl}/${key}`;
  const token = await scry<string>({
    app: 'genuine',
    path: '/secret',
  }).catch((e) => {
    logger.log('failed to get secret', { e });
    return '';
  });
  return `${url}?token=${token}`;
};

export const getFinalMemexUrl = async (memexUploadUrl: string) => {
  const fileUrlResponse = await fetch(memexUploadUrl);
  const fileUrl = await fileUrlResponse.json().catch(() => {
    logger.log('Error parsing response body, fileUrlResponse');
    return '';
  });

  console.log(`final memex url`, fileUrl);

  return fileUrl;
};
