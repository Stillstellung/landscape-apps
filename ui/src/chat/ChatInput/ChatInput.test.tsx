import { describe, it, expect } from 'vitest';
import React from 'react';
import { render } from '../../../test/utils';
import ChatInput from './ChatInput';

vi.mock('@/components/Leap/useLeap', () => ({
  default: () => ({
    isOpen: false,
  }),
}));

vi.mock('@/state/groups', () => ({
  useGroupFlag: () => '~sampel-palnet/test',
  useGroups: () => ({}),
}));

vi.mock('@/logic/useGroupPrivacy', () => ({
  default: () => 'public',
}));

vi.mock('@/logic/analytics', () => ({
  captureGroupsAnalyticsEvent: () => ({}),
}));

describe('ChatInput', () => {
  it.skip('renders as expected', () => {
    const { asFragment } = render(
      <ChatInput
        whom="~zod/test"
        sendDm={() => {
          // placeholder;
        }}
        dropZoneId="dropzone"
        isScrolling={false}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
