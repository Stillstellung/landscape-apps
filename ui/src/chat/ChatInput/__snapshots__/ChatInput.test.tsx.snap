// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatInput > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="flex w-full items-end space-x-2"
  >
    <div
      class="flex-1"
      id="dropzone"
    >
      <div
        class="relative flex items-end justify-end"
      >
        <div
          class="relative flex flex-none items-center justify-center rounded bg-black w-6 h-6 rounded p-1.5 mr-2 mb-1"
          style="background-color: rgb(0, 0, 0);"
        >
          <urbit-sigil
            background="#000000"
            detail="none"
            foreground="white"
            point="~finned-palmer"
            size="12"
            space="none"
          />
        </div>
        <div
          class="input block p-0 w-full break-words"
        >
          <div
            class="w-full"
          >
            <div
              aria-label="Message editor with formatting menu"
              class="ProseMirror input-inner"
              contenteditable="true"
              spellcheck="false"
              tabindex="0"
              translate="no"
            >
              <p
                class="is-empty is-editor-empty"
                data-placeholder="Message"
              >
                <br
                  class="ProseMirror-trailingBreak"
                />
              </p>
            </div>
          </div>
          <div
            class="pointer-events-none fixed"
          />
        </div>
      </div>
    </div>
    <button
      aria-label="Send message"
      class="button px-2"
      disabled=""
    >
      <svg
        class="h-4 w-4"
        fill="none"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="stroke-current"
          d="M8 3L4 7M8 3L12 7M8 3V13"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
